src/devices/machine/mc68328.cpp

```diff
@@ -1265,115 +1701,131 @@ uint8_t mc68328_device::pcsel_r() // 0x413
 
-void mc68328_device::set_port_d_lines(uint8_t state, int bit)
+void mc68328_base_device::port_d_in_w(int state, int bit)
 {
-	uint8_t old_button_state = m_pddata;
+	const u8 old_state = m_pdindata;
+	m_pdindata &= ~(1 << bit);
+	m_pdindata |= state << bit;
 
-	if (state & (1 << bit))
+	// If no bit has changed state, there's nothing to do.
+	if (old_state == m_pdindata)
 	{
-		m_pddata |= (1 << bit);
-	}
-	else
-	{
-		m_pddata &= ~(1 << bit);
+		return;
 	}
 
-	m_pddataedge |= ~old_button_state & m_pddata;
-
-	poll_port_d_interrupts();
-}
-
-void mc68328_device::poll_port_d_interrupts()
-{
-	uint8_t line_transitions = m_pddataedge & m_pdirqedge;
-	uint8_t line_holds = m_pddata &~ m_pdirqedge;
-	uint8_t line_interrupts = (line_transitions | line_holds) & m_pdirqen;
-
-	if (line_interrupts)
-	{
-		set_interrupt_line(line_interrupts << 8, 1);
-	}
-	else
+	// If we're not edge-triggered, handle potential level-sensitive interrupts.
+	if (!BIT(m_pdirqedge, bit))
 	{
-		set_interrupt_line(INT_KBDINTS, 0);
+		// If the new state is low while PDPOL is active-low (0) or vice-versa, assert the interrupt for this bit.
+		// Otherwise, clear it.
+		set_interrupt_line(INT_KBDINTS + bit, (int)(state == BIT(m_pdpol, bit)));
+		return;
 	}
+
+	set_interrupt_line(INT_KBDINTS + bit, 1);
 }
 
-void mc68328_device::pddir_w(uint8_t data) // 0x418
+void mc68328_base_device::pddir_w(u8 data) // 0x418
 {
-	LOGMASKED(LOG_GPIO_D, "pddir_w: PDDIR = %02x\n", data);
+	LOGMASKED(LOG_GPIO_D, "%s: pddir_w: PDDIR = %02x\n", machine().describe_context(), data);
 	m_pddir = data;
 }
 
-uint8_t mc68328_device::pddir_r() // 0x418
+u8 mc68328_base_device::pddir_r() // 0x418
 {
-	LOGMASKED(LOG_GPIO_D, "pddir_r: PDDIR: %02x\n", m_pddir);
+	LOGMASKED(LOG_GPIO_D, "%s: pddir_r: PDDIR: %02x\n", machine().describe_context(), m_pddir);
 	return m_pddir;
 }
 
-void mc68328_device::pddata_w(uint8_t data) // 0x419
+void mc68328_base_device::pddata_w(u8 data) // 0x419
 {
-	LOGMASKED(LOG_GPIO_D, "pddata_w: PDDATA = %02x\n", data);
+	LOGMASKED(LOG_GPIO_D, "%s: pddata_w: PDDATA = %02x (outputing %02x)\n", machine().describe_context(), data, data & m_pddir);
 	m_pddata = data;
-	m_out_port_d_cb(m_pddata);
+
+	for (int bit = 0; bit < 4; bit++)
+	{
+		if (BIT(m_pdirqedge & data, bit))
+		{
+			set_interrupt_line(INT_KBDINTS + bit, 0);
+		}
+	}
+
+	for (int i = 0; i < 8; i++)
+	{
+		if (BIT(m_pddir, i) && !m_out_port_d_cb[i].isnull())
+		{
+			m_out_port_d_cb[i](BIT(m_pddata, i));
+		}
+	}
 }
 
-uint8_t mc68328_device::pddata_r() // 0x419
+u8 mc68328_base_device::pddata_r() // 0x419
 {
-	LOGMASKED(LOG_GPIO_D, "pddata_r: PDDATA: %02x\n", m_pddata);
-	if (!m_in_port_d_cb.isnull())
+	u8 data = 0;
+	for (int i = 0; i < 8; i++)
 	{
-		return m_in_port_d_cb(0);
+		if (BIT(m_pddir, i))
+		{
+			data |= m_pddata & (1 << i);
+		}
+		else if (!m_in_port_d_cb[i].isnull())
+		{
+			data |= m_in_port_d_cb[i]() << i;
+		}
+		else
+		{
+			data |= m_pdpuen & (1 << i);
+		}
```

Comment:
Similarly, is the `isnull` check worth it, or does it go OK if you use `resolve_safe(1)` on the callbacks?